[wingman -> rllib] IMPALA MultiDiscrete changes

bjg2 opened the pull request:
NOTE: This is the beginning of the pull request, so we can align. No unit tests were run, nor the changes are tested broader than we need them.

## What do these changes do?

- Enables the MultiDiscrete actions space in IMPALA.

- Adds the MultiCategorical action distribution - we use IMPALA with that distribution - out model outputs categorical logits for each action. Possibly not all action spaces work for all action distributions, tested is that Discrete works with DiagGaussian (default) and that MultiDiscrete works with categorical (MultiCategorical). Note that model actually outputs a 1-dimensional tensor, that is reshaped to logit sizes according to provided MultiDiscrete action space.

- VTrace is replaced with an implementation that works for MultiDiscrete action space as well.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/11614/
Test FAILed.

ericl commented:
Do you know what's going on with the formatting changes? It makes it kind of hard to review.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/11638/
Test FAILed.

bjg2 commented:
> Do you know what's going on with the formatting changes? It makes it kind of hard to review.

Aligned formatting with yours impala.py, it should be ok to take a look at merge request now.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/11643/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/11644/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/11645/
Test FAILed.

bjg2 commented on python/ray/rllib/agents/impala/impala.py:
```diff
@@ -70,6 +70,9 @@
     "broadcast_interval": 1,
 
+    # Actions are chosen based on this distribution, if provided
+    "dist_type": None,
```
It is, as default is DiagGaussian, and we use Categorical (which is MultiCategorical for MultiDiscrete action space).

bjg2 commented on python/ray/rllib/agents/impala/vtrace_policy_graph.py:
```diff
@@ -101,6 +109,11 @@ def __init__(self,
             "Must use `truncate_episodes` batch mode with V-trace."
         self.config = config
         self.sess = tf.get_default_session()
+        self._is_discrete = False
```
Agree, fixed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/11688/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/11690/
Test FAILed.

pimpke commented on python/ray/rllib/agents/impala/vtrace.py:
```diff
@@ -189,10 +190,8 @@ def from_importance_weights(log_rhos,
-      log(target_policy(a) / behaviour_policy(a)). V-trace performs operations
-      on rhos in log-space for numerical stability.
+    rhos: A float32 tensor of shape [T, B, NUM_ACTIONS] representing the
+      importance sampling weights, i.e. target_policy(a) / behaviour_policy(a).
```
We were experimenting with both log prob and prob spaces, will return it to log space.

pimpke commented on python/ray/rllib/agents/impala/vtrace.py:
```diff
@@ -33,14 +33,14 @@
-    'vs', 'pg_advantages', 'log_rhos', 'behaviour_action_log_probs',
-    'target_action_log_probs'
+    'vs', 'pg_advantages', 'rhos', 'behaviour_action_policy',
+    'target_action_policy'
```
Yup.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/11693/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/11774/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/11775/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/11776/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/11777/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/11947/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/11945/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/11946/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/11948/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/11949/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/11950/
Test FAILed.

bjg2 commented on python/ray/rllib/agents/impala/vtrace.py:
```diff
@@ -33,14 +33,14 @@
 nest = tf.contrib.framework.nest
```
Done.

bjg2 commented on python/ray/rllib/agents/impala/vtrace.py:
```diff
@@ -33,14 +33,14 @@
 
 
-def log_probs_from_logits_and_actions(policy_logits, actions):
+def select_policy_values_using_actions(policy_logits, actions):
```
done.

bjg2 commented on python/ray/rllib/agents/impala/vtrace.py:
```diff
@@ -93,10 +92,10 @@ def from_logits(behaviour_policy_logits,
 
   Args:
-    behaviour_policy_logits: A float32 tensor of shape [T, B, NUM_ACTIONS] with
+    behaviour_policy: A float32 tensor of shape [T, B, NUM_ACTIONS] with
```
done.

ericl commented:
I pushed some changes to allow vtrace_test to run on the modified implementation. Looks like most tests pass but test_vtrace_from_logits_Batch1 and test_vtrace_from_logits_Batch2 don't, could you take a look?

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/12008/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/12010/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/12079/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/12082/
Test FAILed.

bjg2 commented:
> I pushed some changes to allow vtrace_test to run on the modified implementation. Looks like most tests pass but test_vtrace_from_logits_Batch1 and test_vtrace_from_logits_Batch2 don't, could you take a look?

Fixed those tests with https://github.com/ray-project/ray/pull/3967/commits/f161edf89e996ce734ac3aa3fd4cf079f8dfbfe9

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/12088/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/12101/
Test FAILed.

bjg2 commented:
> I'm a bit confused, isn't it Categorical for discrete spaces?
>
> ```
>        elif isinstance(action_space, gym.spaces.Discrete):
>            return Categorical, action_space.n
> ```
> I don't think DiagGaussian ever gets used in IMPALA does it (maybe you meant APPO?)

Good catch - my bad. I was first using Box for IMPALA, as I didn't see MultiDiscrete action space at the moment. For Box, the default was DiagGaussian. When I switched to MultiDiscrete I forgot dist_type was not needed anymore. Fixed with https://github.com/ray-project/ray/pull/3967/commits/d32d253c63ca73944ada0dc887a34826e4f37b6f.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/12115/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/12152/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/12157/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/12158/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/12293/
Test FAILed.

bjg2 commented:
Is this one waiting on something?

ericl commented:
Ah looks like test_supported_spaces is failing on APPO somehow. I can take a look in a bit.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/12325/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/12355/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/12446/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/12447/
Test FAILed.

ericl commented:
Merged, thanks!
